lib/extract.ts

```diff
@@ -0,0 +1,107 @@
+import { promises } from 'fs';
+import { join, extname, basename, relative } from 'path';
+import { FilePath } from './types';
+import { DECOMPRESSING_CONCURRENCY_LEVEL, EXTRACTED_DIR_FLAG } from './common';
+import { debug } from './debug';
+
+const pMap = require('p-map');
+const AdmZip = require('adm-zip');
+const tar = require('tar');
+
+const { mkdir } = promises;
+
+const zipFormats = ['.zip', '.zipx'];
+const tarFormats = ['.tar', '.gz', '.tgz'];
+
+interface ExtractionHandler {
+  (path: FilePath): void;
+}
+
+async function handleExtraction(
+  path: FilePath,
+  temporaryDir: FilePath,
+  childArchiveHandler: ExtractionHandler,
+) {
+  const extractionTarget: FilePath = join(
+    temporaryDir,
+    path.includes(temporaryDir)
+      ? relative(temporaryDir, `${path}${EXTRACTED_DIR_FLAG}`)
```

Comment:
I would rename `EXTRACTED_DIR_FLAG` to `EXTRACTED_DIR_SUFFIX`

Reply:
Yep, good suggestion! I'll update it right now.